### PR TITLE
<a> tag mapping to manage URL #fragment

### LINKED ISSUE #1346: Richtext/HTML in page anchor link
By considering that the standard mapping is thought as a HTML mapping, a missing feature is the management of _in page anchor_ (#something) or  `URL fragment` (see `<a>`  [specs](https://developer.mozilla.org/fr/docs/Web/HTML/Element/a)).

My use case is to import HTML data containing footnotes:
```html
<p>Texte 1<a id="ref_1_note_1" href="#note_1">[1]</a></p> 
...
<p>Texte 2<a id="ref_2_note_1" href="#note_1">[1]</a></p> 
...
<p>Texte 3<a id="ref_1_note_2" href="#note_2">[2]</a></p> 
...
<p id="note_1">1. <a href="#ref_1_note_1">a</a>, <a href="#ref_2_note_1">b</a> La note 1</p>
<p id="note_2">2. <a href="#ref_1_note_2">^</a> La note 2</p>
```

One way is to create my own standoff mapping, but then, users cannot edit anymore their HTML in Salsah (because footnotes won’t be managed by the default CKEditor and because of #908)

An other way is to:

1. Enhanced standard mapping to import these new tags:
     -  `<a id="xxxx" href="#yyyyy">` 
     -  `<p id="yyyy">` 

2. Add footnote edition tools to Salsah by adding this plugin for example: https://ckeditor.com/cke4/addon/footnotes

**To reduce devs** we could limit the footnote to a simple link between two anchors (so you couldn’t add several refs to the same note):
```html
<p>Texte 1<a id="ref_note_1" href="#note_1">[1]</a></p> 
...
<p>Texte 2<a id="ref_note_2" href="#note_2">[2]</a></p> 
...
<p>Texte 3<a id="ref_note_3" href="#note_3">[3]</a></p> 
...
<a id="note_1" href="#ref_note_1">1.</a><p>La note 1</p>
<a id="note_2" href="#ref_note_2">2.</a><p>Voir note <a href="#note_1">1</a></p>
<a id="note_3" href="#ref_note_3">3.</a><p>La note 3</p>
```

1. Enhanced standard mapping to import this new anchor tag:  `<a id="xxxx" href="#yyyyy">` 

2. Enhance `link` button to add an `id` field edition

Some feedback?

### PR DESCRIPTION
**Pull request by gfoo**
resolves #1346

### CODE REVIEW

**Comment by gfoo** [comment]
- [ ]  Add URL #fragment  `<a>` mapping
- [ ]  ttl generation (@loicjaouen  @tobiasschweizer  any docs somewhere?)
- [ ]  Update Knora [doc](https://docs.knora.org/paradox/03-apis/api-v1/xml-to-standoff-mapping.html#description) 
- [ ]  Add test
- [ ] Upgrade script

**Comment by gfoo** [comment]
> @loicjaouen @tobiasschweizer any docs somewhere?

I think I have to:

1. create the `standard-mapping.json`
```json
{
    "knora-api:mappingHasName": "???",
    "knora-api:attachedToProject": {
        "@id": "???"
    },
    "rdfs:label": "???",
    "@context": {
        "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
        "knora-api": "http://api.knora.org/ontology/knora-api/v2#"
    }
}
```

2. call the endpoint to generate the new mapping:
```shell
$ curl -u root@example.com:test -X POST -F json=@standard-mapping.json -F xml=@standard-mapping.xml  http://localhost:3333/v2/mapping
```

3. generate the ttl:
```sparql
PREFIX knora-base: <http://www.knora.org/ontology/knora-base#>
Construct {
    ?mapping ?p ?o .
    ?mapping knora-base:hasMappingElement ?mEle .
    ?mEle ?pp ?oo .
    ?oo ?ppp ?ooo .
}
WHERE {
    BIND(<http://rdfh.ch/standoff/mappings/StandardMapping> as ?mapping)
    ?mapping ?p ?o .

    OPTIONAL {
        ?mapping knora-base:hasMappingElement ?mEle .
    	OPTIONAL {
        	?mEle ?pp ?oo .
    		OPTIONAL {
        		?oo ?ppp ?ooo .
            }
        }
    }
}
```



in fact I know how to call

**Comment by gfoo** [comment]
I'm a little afraid of updating the ttl  using a sparql query because of IRIs... maybe we should update the ttl by hand? :|

And what about the upgrading of an existing  Knora base?

**Comment by tobiasschweizer** [comment]
Yes, I think you have to add the new element to `/Users/tobi/gitreps/Knora_github/webapi/_test_data/test_route/texts/mappingForStandardHTML.xml`, create the new mapping in Knora and update `/Users/tobi/gitreps/Knora_github/knora-ontologies/standoff-data.ttl`.

Maybe you can do this incrementally, otherwise the IRIs will be all different and every `TextValue` with standoff would ave to be updated (its reference to the mapping).

**Comment by tobiasschweizer** [comment]
> I'm a little afraid of updating the ttl using a sparql query because of IRIs.

I am suffering from PTSD: posttraumatic **standoff** disorder myself ..

**Comment by gfoo** [comment]
@tobiasschweizer is there any tests to update?

@tobiasschweizer @benjamingeer As you can see in this commit 380121a I have *just* to add few triplets to an existing Knora base to upgrade it. Is that correct for you? Any guide lines or I just try to copy/paste the way you are doing your upgrade scripts in `upgrade` folder?

**Comment by tobiasschweizer** [comment]
>  is there any tests to update?

You added an element to the existing mapping, so everything is expected to run as before (so the tests should pass). 

I think you should add a test that checks the new element.

**Comment by gfoo** [comment]
> I think you should add a test that checks the new element.

in which test file? (sorry, never used your test framework...)

**Comment by tobiasschweizer** [comment]
> in which test file?

that would be `/Users/tobi/gitreps/Knora_github/webapi/src/test/scala/org/knora/webapi/e2e/v1/StandoffV1R2RSpec.scala`

**Comment by gfoo** [comment]
Not very confortable with the CI failed, any clues?

**Comment by tobiasschweizer** [comment]
https://travis-ci.org/dhlab-basel/Knora/jobs/586477300#L6096

**Comment by gfoo** [comment]
@tobiasschweizer 
I don't understand why the test [`perform a fulltext search for 'Dinge'`](https://travis-ci.org/dhlab-basel/Knora/jobs/586584522#L6133) fails, probably a side effect of my modification of `standoff-data.ttl`, but why? 

It seems that an `internal-link` appears in the result (generating the diff error). If the test data come from the `anything-data.ttl`, there is no such data like that. 

The cited [value](https://github.com/dhlab-basel/Knora/blob/a4d2cff1365f289259802c56065f7f03b529e073/webapi/_test_data/all_data/anything-data.ttl#L604) in the result is a `TextValue` containing a `StandoffLinkTag` but not an internal link...

**Comment by tobiasschweizer** [comment]
> I don't understand why the test perform a fulltext search for 'Dinge' fails, probably a side effect of my modification of standoff-data.ttl, but why?

I think this is because the mapping is returned as well at it contains more elements than before.

**Comment by tobiasschweizer** [comment]
see https://github.com/dhlab-basel/Knora/blob/6555a8e7e4dfeb280c44c3bebc6fabc8ea8fcac3/webapi/src/main/scala/org/knora/webapi/messages/v2/responder/valuemessages/ValueMessagesV2.scala#L1246-L1253

**Comment by gfoo** [comment]
I added a v1 test [there](https://github.com/dhlab-basel/Knora/blob/bfa2750da823d08821a396fa283c662be27248ca/webapi/src/test/scala/org/knora/webapi/e2e/v1/StandoffV1R2RSpec.scala#L920) using this [data](https://github.com/dhlab-basel/Knora/blob/wip/standoff-internal-anchor/webapi/_test_data/test_route/texts/StandardHTML-internal-link.xml) as input, but when I try to send the same data to the v2 API I get an error:
```json
{
  "@type" : "anything:Thing",
  "anything:hasRichtext" : {
    "@type" : "knora-api:TextValue",
    "knora-api:textValueAsXml" : "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<text><p>ref to note <a class=\"internal-link\" href=\"#_note1\" id=\"_ref-note1\">[1]</a></p><p><a class=\"internal-link\" href=\"#_ref-note1\" id=\"_note1\">[1]</a> note 1</p></text>",
    "knora-api:textValueHasMapping" : {
      "@id" : "http://rdfh.ch/standoff/mappings/StandardMapping"
    }
  },
  "knora-api:attachedToProject" : {
    "@id" : "http://rdfh.ch/projects/0001"
  },
  "rdfs:label" : "test thing",
  "@context" : {
    "rdf" : "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
    "knora-api" : "http://api.knora.org/ontology/knora-api/v2#",
    "rdfs" : "http://www.w3.org/2000/01/rdf-schema#",
    "xsd" : "http://www.w3.org/2001/XMLSchema#",
    "anything" : "http://0.0.0.0:3333/ontology/0001/anything/v2#"
  }
}
```

```json
{
    "knora-api:error": "org.knora.webapi.AssertionException: Resource <http://rdfh.ch/0001/JdZmxDDiSvi_OqU7I5Ur0g> was saved, but one or more of its values are not correct",
    "@context": {
        "knora-api": "http://api.knora.org/ontology/knora-api/v2#"
    }
}
```

@benjamingeer it sounds like an encoding problem?

**Comment by gfoo** [comment]
I'm going to try this test [there](https://github.com/dhlab-basel/Knora/blob/bfa2750da823d08821a396fa283c662be27248ca/webapi/src/test/scala/org/knora/webapi/e2e/v2/ValuesRouteV2E2ESpec.scala#L370)

**Comment by gfoo** [comment]
Error is repoduced there: https://travis-ci.org/dhlab-basel/Knora/jobs/586938555#L570

**Comment by gfoo** [comment]
@benjamingeer my test above reveals an error described here #1221 (fixed here #1224)

Don't know if it is related to my first error: `was saved, but one or more of its values are not correct` ?

**Comment by benjamingeer** [comment]
@gfoo Thanks, I will have a look.

**Comment by benjamingeer** [comment]
The test fails because this standoff tag is supposed to be saved:

```
StandoffTagV2(
  endParentIndex = None, 
  originalXMLID = None, 
  uuid = UUID(
    leastSigBits = -5337030549628269419, 
    mostSigBits = -299700061709319665
  ),
  endPosition = 88, 
  startParentIndex = Some(0), 
  attributes = Vector(StandoffTagInternalReferenceAttributeV2(
    standoffPropertyIri = "http://www.knora.org/ontology/knora-base#standoffTagHasInternalReference".toSmartIri, 
    value = "link_id"
  )),
  startIndex = 3, 
  endIndex = None, 
  dataType = Some(http://www.knora.org/ontology/knora-base#StandoffInternalReferenceTag), 
  startPosition = 84, 
  standoffTagClassIri = "http://www.knora.org/ontology/knora-base#StandoffInternalReferenceTag".toSmartIri
)
```

But when we read it back from the triplestore, it looks like this:

```
StandoffTagV2(
  endParentIndex = None, 
  originalXMLID = None, 
  uuid = UUID(
    leastSigBits = -5337030549628269419, 
    mostSigBits = -299700061709319665
  ), 
  endPosition = 88, 
  startParentIndex = Some(0), 
  attributes = Vector(StandoffTagStringAttributeV2(
    standoffPropertyIri = "http://www.knora.org/ontology/knora-base#standoffTagHasInternalReference".toSmartIri, 
    value = "#link_id"
  )), 
  startIndex = 3, 
  endIndex = None, 
  dataType = Some(http://www.knora.org/ontology/knora-base#StandoffInternalReferenceTag), 
  startPosition = 84, 
  standoffTagClassIri = "http://www.knora.org/ontology/knora-base#StandoffInternalReferenceTag".toSmartIri
)
```

So we expect to read a `StandoffTagInternalReferenceAttributeV2`, but instead we read a `StandoffTagStringAttributeV2`. This could mean either that it was written wrong, or that it's read wrong.

**Comment by gfoo** [comment]
did you also noticed the diff value `link_id` and `#link_id`?

**Comment by benjamingeer** [comment]
Yes, I'm trying to figure out what's going on there.

**Comment by benjamingeer** [comment]
@tobiasschweizer I think I've fixed this in ab3adaf2df1816d5b4c636bff9f6b0b9b380861b, but I don't understand why it worked before this PR. Could you please have a look?

The test is `"create a text value with standoff"` in `ValuesRouteV2E2ESpec`. This is the text submitted:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<text>
   This text links to another <a class="salsah-link" href="http://rdfh.ch/0001/another-thing">resource</a>.
   And this <strong id="link_id">strong value</strong> is linked by this <a class="internal-link" href="#link_id">link</a>
</text>
```

The `<strong>` tag is saved like this:

```turtle
<http://rdfh.ch/0001/a-thing/values/SIyqI_D-SMuQb2OZ-UIu_Q/standoff/2> a standoff:StandoffBoldTag ;
  knora-base:standoffTagHasStart "53"^^xsd:integer ;
  knora-base:standoffTagHasEnd "65"^^xsd:integer ;
  knora-base:standoffTagHasStartIndex "2"^^xsd:integer ;
  knora-base:standoffTagHasStartParent <http://rdfh.ch/0001/a-thing/values/SIyqI_D-SMuQb2OZ-UIu_Q/standoff/0> ;
  knora-base:standoffTagHasUUID "wiZYQjR_TwOMM0kzis6tuw" ;
  knora-base:standoffTagHasOriginalXMLID "link_id" .
```

The `<a>` tag is saved like this:

```turtle
<http://rdfh.ch/0001/a-thing/values/SIyqI_D-SMuQb2OZ-UIu_Q/standoff/3> a knora-base:StandoffInternalReferenceTag ;
  knora-base:standoffTagHasInternalReference <http://rdfh.ch/0001/a-thing/values/SIyqI_D-SMuQb2OZ-UIu_Q/standoff/2> ;
  knora-base:standoffTagHasStart "84"^^xsd:integer ;
  knora-base:standoffTagHasEnd "88"^^xsd:integer ;
  knora-base:standoffTagHasStartIndex "3"^^xsd:integer ;
  knora-base:standoffTagHasStartParent <http://rdfh.ch/0001/a-thing/values/SIyqI_D-SMuQb2OZ-UIu_Q/standoff/0> ;
  knora-base:standoffTagHasUUID "-9dAYbqnQg-17w797f_MlQ" .
```

This looks correct to me. It looks to me like the `StandoffInternalReferenceTag ` is read back incorrectly from the triplestore, so it gets a `StandoffTagStringAttributeV2` instead of a `StandoffTagInternalReferenceAttributeV2`. I think the problem is on this line:

https://github.com/dhlab-basel/Knora/blob/develop/webapi/src/main/scala/org/knora/webapi/util/standoff/StandoffTagUtilV2.scala#L786

But that test has been there for a long time. Why didn't it fail before?

**Comment by tobiasschweizer** [comment]
@benjamingeer I think this is because the tests for internal links were using v1 like https://github.com/dhlab-basel/Knora/blob/fb7adba749bf6b37ed36f49f2d6cb3f553b92382/webapi/src/test/scala/org/knora/webapi/e2e/v1/StandoffV1R2RSpec.scala#L669-L697

You add a line to https://github.com/dhlab-basel/Knora/pull/1440/files#diff-85a5990b11ab2dfe887be3905afb24bc to check for internal links which was not there before.

**Comment by tobiasschweizer** [comment]
Probably it was always wrong but v1 did not check for the attribute classes.

**Comment by benjamingeer** [comment]
> You add a line to https://github.com/dhlab-basel/Knora/pull/1440/files#diff-85a5990b11ab2dfe887be3905afb24bc to check for internal links which was not there before.

OK, makes sense! I hadn't noticed that the test had been changed.

**Comment by benjamingeer** [comment]
@tobiasschweizer Also, there was another bug on the same line of code: the `#` character was inserted into the attribute when constructing a `StandoffTagV2` from triplestore query results, but in `StandoffTagUtilV2.convertXMLStandoffTagToStandoffTagV2`, it's clear that the `#` should not be included in a `StandoffTagV2`. Instead, it should be added when generating XML. Fixed in f156202. Does this make sense to you?

**Comment by tobiasschweizer** [comment]
> Does this make sense to you?

I need to look at it more carefully. I will do it this afternoon.

**Comment by tobiasschweizer** [comment]
> Instead, it should be added when generating XML. Fixed in f156202. Does this make sense to you?

Yes, I think it makes sense to add the "#" when converting the standoff attribute to XML.

**Comment by gfoo** [comment]
> Could you add a test case doing standoff attribute conversions to /Users/tobi/gitreps/Knora_github/webapi/src/test/scala/org/knora/webapi/util/standoff/StandoffTagUtilV2Spec.scala?

Sorry, but is this a request for me or for @benjamingeer?

**Comment by gfoo** [comment]
If it helps, tested the branch by importing some data with  internal links and I have errors from apparently `StandoffTagUtilV2`:
```
webapi_1   | 2019-09-24 13:05:53 | ERROR | Dispatcher | key not found: http://rdfh.ch/0113/jk53EvM6THGkYUC5hqHpug/values/uPqkYxHMTpSSKQP9eBNDqA/standoff/100
webapi_1   | java.util.NoSuchElementException: key not found: http://rdfh.ch/0113/jk53EvM6THGkYUC5hqHpug/values/uPqkYxHMTpSSKQP9eBNDqA/standoff/100
webapi_1   | 	at scala.collection.MapLike.default(MapLike.scala:235)
webapi_1   | 	at scala.collection.MapLike.default$(MapLike.scala:234)
webapi_1   | 	at scala.collection.AbstractMap.default(Map.scala:63)
webapi_1   | 	at scala.collection.MapLike.apply(MapLike.scala:144)
webapi_1   | 	at scala.collection.MapLike.apply$(MapLike.scala:143)
webapi_1   | 	at scala.collection.AbstractMap.apply(Map.scala:63)
webapi_1   | 	at org.knora.webapi.util.standoff.StandoffTagUtilV2$.$anonfun$createStandoffTagsV2FromSelectResults$7(StandoffTagUtilV2.scala:783)
webapi_1   | 	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:237)
webapi_1   | 	at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:234)
webapi_1   | 	at scala.collection.TraversableLike.map(TraversableLike.scala:237)
webapi_1   | 	at scala.collection.TraversableLike.map$(TraversableLike.scala:230)
webapi_1   | 	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
webapi_1   | 	at org.knora.webapi.util.standoff.StandoffTagUtilV2$.$anonfun$createStandoffTagsV2FromSelectResults$6(StandoffTagUtilV2.scala:767)
webapi_1   | 	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:237)
webapi_1   | 	at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:234)
webapi_1   | 	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:465)
webapi_1   | 	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:465)
webapi_1   | 	at scala.collection.TraversableLike.map(TraversableLike.scala:237)
webapi_1   | 	at scala.collection.TraversableLike.map$(TraversableLike.scala:230)
webapi_1   | 	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
webapi_1   | 	at org.knora.webapi.util.standoff.StandoffTagUtilV2$.$anonfun$createStandoffTagsV2FromSelectResults$5(StandoffTagUtilV2.scala:749)
webapi_1   | 	at scala.util.Success.$anonfun$map$1(Try.scala:255)
webapi_1   | 	at scala.util.Success.map(Try.scala:213)
webapi_1   | 	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
webapi_1   | 	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
webapi_1   | 	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
webapi_1   | 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
webapi_1   | 	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
webapi_1   | 	at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:91)
webapi_1   | 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
webapi_1   | 	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:85)
webapi_1   | 	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:91)
webapi_1   | 	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
webapi_1   | 	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:44)
webapi_1   | 	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
webapi_1   | 	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
webapi_1   | 	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
webapi_1   | 	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by gfoo** [comment]
@benjamingeer do you mean that this error (related to the paging) is due to the long size of the xml input? If yes, I can add a test into in `ValuesRouteV2E2ESpec` ?

**Comment by tobiasschweizer** [comment]
> That would be great, because I don't understand what you mean by standoff attribute conversions.

What I have in mind is to call the method `createStandoffTagsV2FromSelectResults` from the standoff utility with some mocked data. So we can check it behaves as expected.

**Comment by gfoo** [comment]
> @benjamingeer do you mean that this error (related to the paging) is due to the long size of the xml input? If yes, I can add a test into in `ValuesRouteV2E2ESpec` ?

Reproduced: https://travis-ci.org/dhlab-basel/Knora/jobs/588991665#L697

**Comment by benjamingeer** [comment]
@tobiasschweizer @gfoo To fix this, I changed the SPARQL queries that query standoff tags, so that when they query a `StandoffInternalReferenceTag`, they return an extra triple with a virtual property, `knora-base:targetHasOriginalXMLID`, containing the XML ID of the target.

**Comment by gfoo** [comment]
My own tests work now 👍

**Comment by tobiasschweizer** [comment]
@benjamingeer 

> That would be great, because I don't understand what you mean by standoff attribute conversions.

Please see https://github.com/dhlab-basel/Knora/pull/1440/commits/41475fd7fbace49e194af4649f3d5c54bd49946c. Dies this make sense to you?

**Comment by tobiasschweizer** [comment]
> The SPARQL results for a StandoffInternalReferenceTag should include knora-base:targetHasOriginalXMLID. Try it and see.

I ran a v1 test to create the test data.

**Comment by benjamingeer** [comment]
> I ran a v1 test to create the test data.

OK but your test is in v2. It would make sense to test both cases.

**Comment by tobiasschweizer** [comment]
> OK but your test is in v2. It would make sense to test both cases.

StandoffUtilV2 is used both in v2 and v1. Yes, I will try to produce tests data for v2 too.

File: webapi/src/main/scala/org/knora/webapi/util/standoff/StandoffTagUtilV2.scala [hunk 1]
```diff
@@ -767,23 +767,31 @@ object StandoffTagUtilV2 {
...
-                                if (standoffEntities.standoffPropertyInfoMap(propIri.toSmartIri).isStandoffInternalReferenceProperty) {
+                                if (propDef.isStandoffInternalReferenceProperty) {
                                     // it refers to a standoff node, recreate the original id
 
                                     // value points to another standoff node
-                                    // get this standoff node and access its original id
-                                    val originalId: String = standoffAssertions(value).getOrElse(OntologyConstants.KnoraBase.StandoffTagHasOriginalXMLID, throw InconsistentTriplestoreDataException(s"referred standoff $value node has no original XML id"))
+
+                                    // If a v2 SPARQL template was used, the XML ID is in this standoff tag.
+                                    val originalId: String = standoffTagAssertions.get(OntologyConstants.KnoraBase.TargetHasOriginalXMLID) match {
```
Comment (benjamingeer) [inline webapi/src/main/scala/org/knora/webapi/util/standoff/StandoffTagUtilV2.scala]: This line is used for v2.

File: webapi/src/main/scala/org/knora/webapi/util/standoff/StandoffTagUtilV2.scala [hunk 2]
```diff
@@ -767,23 +767,31 @@ object StandoffTagUtilV2 {
...
+                                        case Some(targetXmlID) => targetXmlID
+
+                                        case None =>
+                                            // If a v1 SPARQL template was used, we have to get the target node and to get its XML ID.
+                                            standoffAssertions(value).getOrElse(OntologyConstants.KnoraBase.StandoffTagHasOriginalXMLID, throw InconsistentTriplestoreDataException(s"referred standoff $value node has no original XML id"))
```
Comment (benjamingeer) [inline webapi/src/main/scala/org/knora/webapi/util/standoff/StandoffTagUtilV2.scala]: This line is used for v1.

**Comment by tobiasschweizer** [comment]
Ok I added tests for v1 and v2 in https://github.com/dhlab-basel/Knora/pull/1440/commits/40be0416f0e8e42bd413b89d378e6dddf0a7e678

Sorry, but I still don't understand the v2 case: Wouldn't `createStandoffTagsV2FromSelectResults` only be called after all pages of standoff have been gathered?

**Comment by tobiasschweizer** [comment]
Or are the pages of `standoffAssertions` converted to `StandoffTagV2` in single steps and combined?

**Comment by benjamingeer** [comment]
The method `createStandoffTagsV2FromSelectResults` is called by `createStandoffTagsV2FromConstructResults`, which is called by `ConstructResponseUtilV2.makeTextValueContentV2` for each page of standoff results. This is how the standoff route returns one page of standoff at a time to the client. I really recommend reading the comments in `ConstructResponseUtilV2.makeTextValueContentV2`.

**Comment by tobiasschweizer** [comment]
Ok, I understand! So v2 needs to read standoff in parts.

> I really recommend reading the comments in ConstructResponseUtilV2.makeTextValueContentV2.

This the last day I can work on this before my holiday and I really try the best to be helpful.

**Comment by benjamingeer** [comment]
> I really try the best to be helpful.

And I'm trying to be helpful by writing comments in the code so you can read them. :)

**Comment by tobiasschweizer** [comment]
> And I'm trying to be helpful by writing comments in the code so you can read them. :)

Normally I would have. But today I *just* wanted to add that additional test ;-) You know how it is ...

**Comment by benjamingeer** [comment]
OK I think that should fix it, let's see if all the tests pass.

**Comment by tobiasschweizer** [comment]
> OK I think that should fix it, let's see if all the tests pass.

great, thx.

**Comment by benjamingeer** [comment]
@tobiasschweizer Can we merge this now?

**Comment by tobiasschweizer** [comment]
@benjamingeer Yes, please go ahead.

**Comment by tobiasschweizer** [comment]
or as we say in Swiss German: "möööörtsch"

**Comment by benjamingeer** [comment]
@tobiasschweizer Thanks for your help, and have a good holiday!

**Comment by gfoo** [comment]
what a triplet team... ah ah ah :^)

**Pull request merged**